README.md
# FinalProject
## Overview
The project involves utalize equiment that uses scientific computations in order to create frequency graphs of a user-inputed .wav files. This is acheievd by the implimentation of a user-freindly GUI with a load button to upload the file. After a user inputs their .wav file or equivilant, the program will then display six unique grpahs. Theses graphs include waveform, low frequency, high frequency, mid frequency, a combined graph, and spectogram. This is achieved through using API and other Python library tools. 

## Purpose
The purpose of the project is to achievie a functional GUI that can accurately read a wav file and generate frequency graphs corrasponding with the inputed data while working with teammates. The project demonstrated how useful teamwork and colaboration is in modeling a real-world coding scenario. 

## Basic Usage Instructions
Make sure you have Pycharm or Python equivlant installed.
Check if you have the packages listed in requirements.txt


CHANGELOG.md
# Change Log


## Version 1.00 12/07/2023

### Added
Feature FileButton: contains GUI and calls functions for graphing and calculations
Feature FetchFile: contains functions that graphs and calculates

### Bug Fix
Feature FetchFileButton: moved to appropriate location

## Version 2.00 12/08/2023

### Added
Feature checkStereo: converts the data to mono
Feature checkType: converts to a wav file
Featue checkMetaData: checks and prints data associated with the file

### Update
Feature GUIButton: Update GUI Button to have prior functions

## Version 3.00 12/09/2023

### Added
Feature fileDisplay: List box that lists the data for the file
Feature makeWavArray: Opens the wav file and turns it into an array
Feature graphingTools: contains all the functions that grpah, calculate RT60
Feature plotWavForm: plots the frequency of the file
Feature plotSpectogram: plots the spectogram
Feature plotRT60: plots the RT60

### Bug Fix
Feature plotRT60: correct the plot of RT60 using the given code

## Version 4.00 12/10/2023

### Added
Feature mergeButton: merges RT60 graphs together
Feature statsButton: plots the spectogram and displays statistics in the list box
Feature combindedPlots: combines all the plots into one

### Bug Fix
Feature FFMPIG: properly installed and implimented




teset.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile

def find_target_frequency(freq_band):
    for x in freqs:
        if x > freq_band:
            break
    return x

def frequency_check(freq_band):
    target_frequency = find_target_frequency(freq_band)
    index_of_frequency = np.where(freqs == target_frequency)[0][0]
    data_for_frequency = spectrum[index_of_frequency]
    data_in_db_fun = 10 * np.log10(data_for_frequency)
    return data_in_db_fun

def find_nearest_value(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return array[idx]

def plotrt60(file_path, freq_band, freq_name):
    sample_rate, data = wavfile.read(file_path)
    global spectrum, freqs, t, im
    spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
    plt.close()

    data_in_db = frequency_check(freq_band)
    plt.figure()
    plt.title(freq_name)
    plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color ='#004bc6')
    plt.xlabel('Time (s)')
    plt.ylabel('Power (dB)')

    index_of_max = np.argmax(data_in_db)
    value_of_max = data_in_db[index_of_max]
    plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')

    sliced_array = data_in_db[index_of_max:]

    value_of_max_less_5 = value_of_max - 5



    value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
    index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)
    plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

    value_of_max_less_25 = value_of_max - 25
    value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
    index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

    plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')
    rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

    rt60 = rt20 * 3

    plt.grid()
    plt.show()

    return rt60

wavTools.py
import os.path
from pydub import AudioSegment
from scipy.io import wavfile
import numpy as np

def checkType(file):    #checks file extension and changes it if needed
    fileExtension = os.path.splitext(file)[1]
    if fileExtension == ".WAV":
        return
    elif fileExtension == ".mp3":
        temp = AudioSegment.from_mp3(file)
        temp.export(file, format='wav')
        print(temp)
        return
    elif fileExtension == ".aac":
        temp = AudioSegment.from_file(file, format='aac') #needs to be changed to aac not mp3
        temp.export(file, format='wav')
        print(temp)
        return
    else:
        print("Incompatible file type")
        return

def checkMetaData(file):    #prints out metadata related to the sound file
    audio = AudioSegment.from_file(file)

    print("Audio Info:")
    print(f"Channels: {audio.channels}")
    print(f"Sample Width: {audio.sample_width} bytes")
    print(f"Frame Rate: {audio.frame_rate} Hz")
    print(f"Frame Width: {audio.frame_width} bytes")
    print(f"Length: {len(audio)} milliseconds")
    return len(audio)

def checkStereo(file):  #converts to monophonic
    raw_audio = AudioSegment.from_file(file, format="wav")
    mono_audio = raw_audio.set_channels(1)
    mono_audio.export(file, format="wav")

def makeWavArray(file_path):    #used to obtain an array of values to plot waveform
    data = wavfile.read(file_path)
    data = np.array(data[1], dtype=np.float32)
    normalized_data = data / np.max(np.abs(data), axis=0)
    return normalized_data

fileButton.py
import tkinter as tk
from tkinter.filedialog import askopenfile
import os
from FinalProject.wavTools import checkType, checkStereo, checkMetaData, makeWavArray
from FinalProject.graphingTools import plotWaveform, plotSpectogram, plotrt60, combinedPlots
from pydub import AudioSegment


def fetchFile():    #defining button command to open file explorer and then display selected file
    file = askopenfile(mode = 'r')
    if file:
        global file_path
        file_path = file.name   #need file_path for merging graphs
        file_name = os.path.basename(file_path)
        fileDisplay.insert(tk.END, file_name)
        checkType(file_path)    #converting to wav
        checkStereo(file_path)  #converting to mono
        length = checkMetaData(file_path)    #printing file's data and returns wav length
        lengthSec = length / 1000 #converting from ms to seconds

        fileDisplay.insert(tk.END, f"File length: {lengthSec} seconds") #displaying length in listbox

        wavArray = makeWavArray(file_path) #making array of the file's data
        print(wavArray)
        plotWaveform(length, wavArray)

        #plotSpectogram(file_path)   #plotting spectogram

        rt1, max1, test1 = plotrt60(file_path, 250, "Low Frequency")   #plotting and getting max frequency and RT60 values
        rt2, max2, test2 = plotrt60(file_path, 1000, "Mid Frequency")
        rt3, max3, test3 = plotrt60(file_path, 10000, "High Frequency")

        rt1 = abs(rt1)  #making RT60's positive
        rt2 = abs(rt2)
        rt3 = abs(rt3)
        rtdif = (rt1 + rt2 + rt3) / 3
        rtdif -= 0.5
        fileDisplay.insert(tk.END, f"RT60 Difference: {rtdif:.3f} seconds")

        fileDisplay.insert(tk.END, f"Maximum Frequency: {max(max1, max2, max3):.3f} kHz")

        fileDisplay.insert(tk.END, f"Low Frequency RT60: {rt1:.3f} seconds")    #displaying RT60 in listbox
        fileDisplay.insert(tk.END, f"Mid Frequency RT60: {rt2:.3f} seconds")
        fileDisplay.insert(tk.END, f"High Frequency RT60: {rt3:.3f} seconds")

        file.close()

def mergeGraphs():
    if file_path:
        colors = ['#004bc6', '#c6000a', '#6a00c6']
        combinedPlots(file_path, colors)

def displayStats():
    if file_path:
        audio = AudioSegment.from_file(file_path)
        fileDisplay.insert(tk.END, f"Frame Rate: {audio.frame_rate} Hz")
        fileDisplay.insert(tk.END, f"Channels: {audio.channels}")
        fileDisplay.insert(tk.END, f"Sample Width: {audio.sample_width} bytes")
        fileDisplay.insert(tk.END, f"Frame Width: {audio.frame_width} bytes")

        plotSpectogram(file_path)

#GUI
root = tk.Tk()
root.geometry('569x120')
root.title('Sound Modeler    \m/')
button = tk.Button(root, text='Select File', width = 25, command = fetchFile)   #button to display initial graphs
button.grid(row=0, column = 0)

colors = ['#004bc6', '#c6000a', '#6a00c6']
mergeButton = tk.Button(root, text='Merge Graphs' , width=25, command=mergeGraphs) #button to combine graphs
mergeButton.grid(row=0, column=1)

statsButton = tk.Button(root, text='Display Statistics', width=25, command=displayStats)
statsButton.grid(row=0, column=2)

fileDisplay = tk.Listbox(root, height=5, width= 90) #listbox for stats
fileDisplay.grid(row=1, column = 0, columnspan=3)

scrollbar = tk.Scrollbar(root, orient=tk.VERTICAL, command=fileDisplay.yview)   #scrollbar for listbox
scrollbar.grid(row=1, column=3, sticky=tk.N + tk.S)
fileDisplay.config(yscrollcommand=scrollbar.set)

root.mainloop()

graphingTools.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.io import wavfile

def plotWaveform(length, wavArray):
    time = np.linspace(0., length, wavArray.shape[0])
    plt.plot(time, wavArray)
    plt.title('Waveform')
    plt.xlabel("Time [s]")
    plt.ylabel("Amplitude")
    plt.show()

def plotSpectogram(filePath):
    sample_rate, data = wavfile.read(filePath)
    spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
    cbar = plt.colorbar(im)
    plt.title('Spectogram')
    plt.xlabel('Time (s)')
    plt.ylabel('Frequency (Hz)')
    cbar.set_label('Intensity (dB)')
    plt.show()
    return spectrum

def find_target_frequency(freq_band):   #finding value of filtered frequency
    for x in freqs:
        if x > freq_band:
            break
    return x

def frequency_check(freq_band):
    target_frequency = find_target_frequency(freq_band)
    index_of_frequency = np.where(freqs == target_frequency)[0][0]
    data_for_frequency = spectrum[index_of_frequency]
    data_in_db_fun = 10 * np.log10(data_for_frequency)
    return data_in_db_fun

def find_nearest_value(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return array[idx]

def plotrt60(file_path, freq_band, freq_name):
    sample_rate, data = wavfile.read(file_path)
    global spectrum, freqs, t, im
    spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
    plt.close() #closing plot so spectogram isn't replotted

    data_in_db = frequency_check(freq_band)
    plt.figure()
    plt.title(freq_name)
    plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color ='#004bc6')
    plt.xlabel('Time (s)')
    plt.ylabel('Power (dB)')

    index_of_max = np.argmax(data_in_db)
    value_of_max = data_in_db[index_of_max]
    plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')

    sliced_array = data_in_db[index_of_max:]

    value_of_max_less_5 = value_of_max - 5

    value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
    index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)
    plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

    value_of_max_less_25 = value_of_max - 25
    value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
    index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

    plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')
    rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

    rt60 = rt20 * 3

    plt.grid()
    plt.show()

    return rt60, value_of_max_less_25, index_of_max


def combinedPlots(file_path, colors):
    sample_rate, data = wavfile.read(file_path)

    spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
    plt.close()

    plt.figure()
    plt.title("Combined RT60's")
    plt.xlabel('Time (s)')
    plt.ylabel('Power (dB)')

    freq_bands = [250, 1000, 10000]
    for x in range(3):
        data_in_db = frequency_check(freq_bands[x])
        plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color=colors[x])
        index_of_max = np.argmax(data_in_db)
        value_of_max = data_in_db[index_of_max]
        plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')

        sliced_array = data_in_db[index_of_max:]

        value_of_max_less_5 = value_of_max - 5

        value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
        index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)
        plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

        value_of_max_less_25 = value_of_max - 25
        value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
        index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

        plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')
        #rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

        #rt60 = rt20 * 3

    plt.grid()
    plt.show()